### ISSUE TITLE
TMC2209 and Pi with UART

### ISSUE THREAD

**Issue by jasonmnemonic** [teemuatlut/TMCStepper#169]
Hi Folks!

Good work on these drivers! Keep em coming :)

This is probably a question for teemuatlut or wolfgangmauer. I am trying to understand how to use the TMC2209 via UART. So far, I have a setup where I can just simply connect the TMC2209 to my Pi and drive STEP and DIR to move motor and change direction.

However, with the code example from simple.ino (Arduino) which I have adapted to my Pi code, I am unable to use the TMC2209 to change direction via 

```
bool shaft = false;

for ( int = 10000; i >= 0; --i )
{
  // Spam to STEP pin a train of pulses...
}

shaft = !shaft;
driver.shaft(shaft);
```

My test connection is always showing `test_connection: 0` which means connected? 

My linux serial port initialization is as follows:

```
	serial->begin( 9600 ); // also tried with 115200

	driver.begin();                 //  SPI: Init CS pins and possible SW SPI pins
                                  	// UART: Init SW UART (if selected) with default 115200 baudrate
	driver.toff(3);                 // Enables driver in software
	driver.rms_current(400);        // Set motor RMS current
	driver.microsteps(16);          // Set microsteps to 1/16th

	//driver.en_pwm_mode(true);       // Toggle stealthChop on TMC2130/2160/5130/5160
	//driver.en_spreadCycle(false);   // Toggle spreadCycle on TMC2208/2209/2224
	driver.pwm_autoscale(true);     // Needed for stealthChop
```

and the serial code is as follows which I am using is as follows (I am not sure what is not working so decided to use 8 data bit, no parity and 1 stop bit):

```
					struct termios options;

					fd = open(port, flags);
					if (fd == -1) {
						printf("[ERROR] UART open(%s)\n", port);
						return;
					}
					fcntl(fd, F_SETFL, O_RDWR);

					tcgetattr(fd, &options);
					options.c_cflag = CS8 | CLOCAL | CREAD | CBAUDEX;
					options.c_cflag &= ~CBAUD;
					options.c_cflag &= ~PARENB;
					options.c_cflag &= ~CSTOPB;
					options.c_iflag = IGNPAR;
					options.c_oflag = baud;
					options.c_lflag = baud;
					tcflush(fd, TCIFLUSH);
					tcsetattr(fd, TCSANOW, &options);
```

1) Can anyone help me figure out how to send through serial to change direction (shaft) please?

2) Could it be my code where all the `SET_REG` are not working as intended? I do not have any build errors but my GCONF.cpp for `#define SET_REG(SETTING) GCONF_register.SETTING = B; write(GCONF_register.address, GCONF_register.sr)` shows the write maps to linux's unistd.h's ``extern ssize_t write (int __fd, const void *__buf, size_t __n) __wur;` that takes 3 parameters.

Basically, is my setup is incorrect so nothing turns happens? 

3) Also on the TMC IDE; I can set motor velocity or RPM to 2000. What is the equivalent API call in this library to do this?

Thanks.

**Comment by jasonmnemonic** [comment]
OK never mind, to answer my own questions. After some tinkering, the TMC2209 starts to work as described with the code example above. The only correction there is the initialise and configuration of the Pi UART serial port to be 8 databit, no parity and 1 stop bit.

So my answers to my own questions:

1) Starts to work now as described above.

2) I see where the write() comes from; it is from the TMC2208Stepper class.

3) I am still trying to figure this out.

**Comment by jasonmnemonic** [comment]
Yay. I think I figured out everything now.

1. Done
2. Done
3. I think done too! It is VACTUAL()

Now I can go figure how to make it work properly.

Only one comment about this so far. The microsteps( 0 ) probably should be microsteps( 1 ) for fullstep because it is more natural to think that 1 is a fullstep and 2 is half step and 8 is 1/8 -th step and so on.

Keep the good work up and I will be happy to also contribute! Thanks :-)
